### PR TITLE
[fix]: [CDS-82263] : adding support for remote multi env/service

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Describe your changes

Passed the git branch in serviceStepParameters,
Though setting it in matrixMetdataMap
And setting ServicYamlV2 = <+matrix.serviceBranch>
EnvironmentYamlV2.gitBranch = <+matrix.envBranch>

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54688)

### CODE REVIEW

**Comment by tchaurasiya** [comment]
trigger smartchecks

**Comment by tchaurasiya** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 1]
```diff
@@ -761,22 +766,46 @@ private ServicePartResponse executeServicePart(
...
+    }
+    return envGitBranch;
+  }
+
+  @Nullable
+  private static String getServiceGitBranch(ServiceStepV3Parameters stepParameters) {
+    String envGitBranch;
+    if (isBlank(stepParameters.getServiceGitBranch())
+        || stepParameters.getServiceGitBranch().equals(SERVICE_GIT_BRANCH_EXPRESSION)) {
+      envGitBranch = null;
```
Comment (lovish1999) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:802]: :nit:  serviceGitBranch

Comment (tchaurasiya) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:802]: Thanks @lovish1999 , fixed.

**Comment by tchaurasiya** [comment]
trigger uts

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 126 at r2](https://reviewable.io/reviews/harness/harness-core/54688#-NhkR603B2GUqb8LqXcD:-NhkR6047pTnyc8R3_ww:bmogtvc) ([raw file](https://github.com/harness/harness-core/blob/6a2acad5dc2adb87340880470b40dc4121a5da75/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L126)):*
> ```Java
> import lombok.extern.slf4j.Slf4j;
> import org.apache.commons.lang3.StringUtils;
> import org.jetbrains.annotations.Nullable;
> ```

do we need to use Nullable from Jetbrains ?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 789 at r2](https://reviewable.io/reviews/harness/harness-core/54688#-NhkRCqu9NUO4uaO33sz:-NhkRCqu9NUO4uaO33t-:br7s1s3) ([raw file](https://github.com/harness/harness-core/blob/6a2acad5dc2adb87340880470b40dc4121a5da75/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L789)):*
> ```Java
>     String envGitBranch;
>     if (isBlank(stepParameters.getEnvGitBranch())
>         || stepParameters.getEnvGitBranch().equals(ENV_GIT_BRANCH_EXPRESSION)) {
> ```

nit: constant on left side of comparison

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 801 at r2](https://reviewable.io/reviews/harness/harness-core/54688#-NhkRGSe4XpWqVHCNNuc:-NhkRGSe4XpWqVHCNNud:br7s1s3) ([raw file](https://github.com/harness/harness-core/blob/6a2acad5dc2adb87340880470b40dc4121a5da75/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L801)):*
> ```Java
>     String serviceGitBranch;
>     if (isBlank(stepParameters.getServiceGitBranch())
>         || stepParameters.getServiceGitBranch().equals(SERVICE_GIT_BRANCH_EXPRESSION)) {
> ```

nit: constant on left side of comparison

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 798 at r2](https://reviewable.io/reviews/harness/harness-core/54688#-NhkRL8-ER3m9c1GyfmO:-NhkRL8-ER3m9c1GyfmP:b-aa3ws1) ([raw file](https://github.com/harness/harness-core/blob/6a2acad5dc2adb87340880470b40dc4121a5da75/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L798)):*
> ```Java
> 
>   @Nullable
>   private static String getServiceGitBranch(ServiceStepV3Parameters stepParameters) {
> ```

does this need to be static ?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 789 at r2](https://reviewable.io/reviews/harness/harness-core/54688#-NhkRTZk9B0dvmZB4Khk:-NhkRTZk9B0dvmZB4Khl:b8lcgah) ([raw file](https://github.com/harness/harness-core/blob/6a2acad5dc2adb87340880470b40dc4121a5da75/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L789)):*
> ```Java
>     String envGitBranch;
>     if (isBlank(stepParameters.getEnvGitBranch())
>         || stepParameters.getEnvGitBranch().equals(ENV_GIT_BRANCH_EXPRESSION)) {
> ```

will it every be <+matrix.envGitbranch> here, should it not be resolved before during step parameter resolution ?

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54688)* status: 3 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steps/MultiDeploymentSpawnerUtils.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/54688#-NhkRd5L7GZXb8VewX1z:-NhkRd5L7GZXb8VewX2-:bj0k5ig) ([raw file](https://github.com/harness/harness-core/blob/6a2acad5dc2adb87340880470b40dc4121a5da75/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steps/MultiDeploymentSpawnerUtils.java#L42)):*
> ```Java
>   private static final String ENV_GIT_BRANCH = "envGitBranch";
>   public static final String SERVICE_GIT_BRANCH_EXPRESSION = "<+matrix." + SERVICE_GIT_BRANCH + ">";
>   public static final String ENV_GIT_BRANCH_EXPRESSION = "<+matrix." + ENV_GIT_BRANCH + ">";
> ```

similar expressions are also defined in ServiceStepV3

**Comment by tchaurasiya** [comment]
> ```java
> String getServiceGitBranch(ServiceStepV3Par
> ```

as git branch is optional field, if not given in pipeline, this expression will come as it is.

**Comment by tchaurasiya** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54688)* status: 4 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by lovish1999 — APPROVED** [review]

**Comment by lovish1999** [comment]
trigger review

**Pull request merged**